[Windows] Setting for high precision processing / HDR tone mapping

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Decoupled the intermediate target (called IT in the rest of the description) format from the backbuffer format.

The render methods can now choose the IT format, each render method can provide its own logic due to their different constraints. Fallback on backbuffer format when the render method does not choose.

For compatibility with previous behavior, the logic was coded so that it will only "upgrade" the format, the back buffer format will still be used the rest of the time and follow the  "Use 10 bit for SDR" setting.

A setting "High precision processing" was created for constrained system to use the previous behavior and always follow the backbuffer format.

* Pixel shaders format selection: rather straightforward, just check usage compatibility. float16 is preferred due to yuv2rgb shader converting as float16. However that's a small change compared with current usage of RGB10 in HDR situations.

* DVXA is more complicated: the dxva processor and its enumerator are needed for texture format compatibility testing, but opening the dxva processor requires knowledge of the texture format. > catch 22...
  * Resolved by delaying the test of CheckFormats a bit (moved into SetOutputFormat()), so the processor can be Open() without texture format.
  * Test compatibility with both IsFormatSupported (windows 8.1) and IsFormatConversionSupported (Windows 10+)
  * Shouldn't interfere with dxva super resolution scaler, limited to SDR media (no hw to confirm through)

Changes left for later:
* Software renderer has other, bigger problems with HDR (color space...), so not changed here
* Allows refactoring of PR #23336 to directly set the intermediate target format as 8 bits instead of modifying the backbuffer + recreating swap chain, with cascade into the IT format.
* One TODO left in CProcessorHD::IsBT2020Supported() still using backbuffer format - doesn't need to be done now because the results are the same. The static vs member functions in CProcessorHD make things difficult without duplicating large amounts of code. To be taken care of in future refactor.

~~* display IT format in debug OSD (only in log for now). The data available in the renderer but should be displayed in rendersystem section, Not sure what to do.~~ Done

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Users selecting "Auto" with AMD GPU or "Off" value in the "Use 10 bit for SDR" setting get slightly reduced precision on HDR to SDR conversion and tone mapping, with lower resulting SDR picture quality.

The output of the color conversion to RGB is truncated to 8 bits instead of preserving the available 10 bits precision.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Windows 10, AMD GPU, Intel gen 4 and gen 8
Setting "Use 10 bit for SDR" auto/on/off.
Setting "Use limited color range" on/off

PQ and HLG material always get 10 bits (DXVA) or float16 (Pixel Shaders).
SDR media is unchanged (8 or 10 bits precision, following existing logic around "Use 10 bit for SDR")

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
Improved quality of tone mapping output for AMD users and users with "Use 10 bit for SDR" off (slightly reduced banding)

## Screenshots (if appropriate):
Not sure how well you'll be able to see, with lossy encoding of screenshots, scaling, etc...
The difference was a little easier to see in the rotating color patterns in movement when playing, for some reason.

There is supposed to be banding in this sample, but tone mapping with 8 bit data adds another level of banding. Zoom into the middle pattern.

Before - tone mapping with 8 bits data 
![image](https://github.com/xbmc/xbmc/assets/489377/30728989-3782-4794-ab59-f35175653ba6)


After - tone mapping with 10 bits data
![image](https://github.com/xbmc/xbmc/assets/489377/019df09c-3a15-4bec-9142-0b866b7c0ab2)
"smoother" banding

Setting:
![image](https://github.com/xbmc/xbmc/assets/489377/081e2452-d41f-46ec-b463-1fd1f505d6e2)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP committed:
[Windows] Allow the render methods to override the intermediate target format

thexai commented:
My request here is make this optional with a new GUI setting or advanced setting instead of change current behavior.

The reason is that use of `DXGI_FORMAT_R16G16B16A16_FLOAT` pixel format may have big performance penalty on some systems as doubles data usage from 32 bits to 64 bits per pixel.

Note that `DXGI_FORMAT_R10G10B10A2_UNORM` uses 32 bits same as `DXGI_FORMAT_R8G8B8A8_UNORM`.

From Microsoft:

> **Option 1. Use FP16 pixel format and scRGB color space**
> 
> However, that option consumes 64 bits per pixel, which doubles GPU bandwidth and memory consumption compared to the traditional UINT8 pixel formats. In addition, scRGB uses numeric values that are outside the normalized [0, 1] range to represent colors that are outside the sRGB gamut and/or greater than 80 nits of luminance. For example, scRGB (1.0, 1.0, 1.0) encodes the standard D65 white at 80 nits; but scRGB (12.5, 12.5, 12.5) encodes the same D65 white at a much brighter 1000 nits. Some graphics operations require a normalized numeric range, and you must either modify the operation, or re-normalize color values.
> 
> **Option 2: Use UINT10/RGB10 pixel format and HDR10/BT.2100 color space**
> 
> But if your app qualifies for option 2, then that might provide better performance if your app is consuming HDR10-encoded content, such as a video player, or if it mainly will be used in fullscreen scenarios, such as a game. 
> 
> This option consumes the same 32 bits per pixel as traditional UINT8 SDR pixel formats. In addition, on certain GPUs this eliminates some processing needed to convert the content to the HDR10 wire format.

https://learn.microsoft.com/en-us/windows/win32/direct3darticles/high-dynamic-range#option-1-use-fp16-pixel-format-and-scrgb-color-space


Incompatibilities could also arise, e.g. on Xbox. If there is a setting, affected users can disable it. As it is now uses FLOAT 16 as long Pixel Shaders are used and the source is 10 bit (even if setting "Use 10 bit for SDR" set in Never). Maybe a new setting in Player / Processing.

CrystalP commented:
The MS article paragraph is about swap chain back buffers format and HDR output, not exactly the same thing.

In general, as much precision as possible should be kept during processing, and the last copy to back buffer is where dither / quantize / CMS / etc happen to adapt the output to the best format for direct scanout / composition by DWM.
 
For now I wouldn't mind DXGI_FORMAT_R10G10B10A2_UNORM as default choice for Pixel Shaders (already an improvement over 8 bits)

Later a "performance level" setting would be good, with for example low = RGB8, normal = RGB10, high = FP16, placebo = FP32
It could drive precision of the HQ scalers at the same time, and other future processing.

Does that work for you?


Yes would be good to display the render method as well.

thexai commented:
> Later a "performance level" setting would be good, with for example low = RGB8, normal = RGB10, high = FP16, placebo = FP32
> It could drive precision of the HQ scalers at the same time, and other future processing.

Probably FP32 is over sized and not has much sense. 

I'm thinking a much simpler ON/OFF setting "Allow high precision texture for processing" or something similar.

OFF ---> Current behavior: same texture type for intermediate target and backbuffer.
ON ---> New behavior of this PR: allow more precise texture for IT/processing.

This setting can be ON by default.

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/windows/RendererBase.cpp:
```diff
@@ -335,21 +335,28 @@ void CRendererBase::DeleteRenderBuffer(int index)
 {
-  DXGI_FORMAT format = DX::Windowing()->GetBackBuffer().GetFormat();
+  // No format specified by renderer > mirror swap chain's backbuffer format
+  if (format == DXGI_FORMAT_UNKNOWN)
```
This setting can be here:

` if (format == DXGI_FORMAT_UNKNOWN || HighPrecisionTexturesSetting == OFF)`

CrystalP commented:
I'm putting this on hold for a bit, with super resolution and others there are more places using backbuffer as default and this is getting complicated.

So first I'm going to do a separate PR to refactor the processor code, remove code duplication and simplify the situation. No functionality change, but a helpful stepping stone.

Then I'll rebase this.

CrystalP committed:
[Windows] Preserve 10 bits of precision for HDR source ( Pixel Shaders)

CrystalP committed:
[Windows] Add high precision processing setting

CrystalP committed:
[Windows] Show the current intermediate target format in the video dbug osd

CrystalP committed:
[Windows] Preserve 10 bits of precision for HDR source (DXVA)

todo: processor capabilities probing does not use the intermediate target format selection - with current architecture, probing happens once before (GetWeight) and once after selection (Configure). Refactor needed to integrate both.

CrystalP commented:
Updated and took the previous points in consideration (setting, video debug osd).
For the osd I shortened the format names to make room for the super resolution and future messages.

This is still not 100% correct for dxva because of the assumption that both RGB8 and RGB10 are supported for a chosen output color space. It's true for the current known dxva processors, is likely to continue to be true, but there is no guarantee.
For older hardware, an Intel gen 4 is fine but I don't have others or older to test the assumption.

With a bit of ugliness it is possible to have Configure() and ProbeProcessorCaps() evaluate support for the chosen intermediate texture format and report 100% correct information, but it's complicated to get GetWeight() to do the same (the most important function because it triggers fallback when support is missing).

Looks like another refactor is needed to decouple things some more, or is it good enough for now?

@thexai, @afedchin thoughts?

thexai reviewed (commented):
Changes looks fine to me

CrystalP commented on xbmc/cores/VideoPlayer/VideoRenderers/windows/RendererDXVA.h:
```diff
@@ -48,8 +48,10 @@ class CRendererDXVA : public CRendererHQ
+  DXGI_FORMAT CalcIntermediateTargetFormat(const VideoPicture& picture) const;
 
   std::unique_ptr<DXVA::CProcessorHD> m_processor;
+  DXGI_FORMAT m_intermediateTargetFormat;
```
right

CrystalP commented:
Wiki page https://kodi.wiki/view/Settings/Player/Videos#Platform_Specific_settings updated, I added the video super resolution setting while I was at it.
